File: VSCode/ebola_compartmental.R
## PROMPT: Code up a compartmental model for the disease spread of Ebola using R in this file 
## INTERFACE: VS Code

# Ebola compartmental model in R
# This script implements a simple SEIRD model:
# S: susceptible
# E: exposed
# I: infectious
# R: recovered
# D: deceased

if (!requireNamespace("deSolve", quietly = TRUE)) {
  install.packages("deSolve")
}

library(deSolve)

ebola_model <- function(t, state, parameters) {
  with(as.list(c(state, parameters)), {
    N <- S + E + I + R + D
    beta <- parameters["beta"]
    sigma <- parameters["sigma"]
    gamma <- parameters["gamma"]
    mu <- parameters["mu"]

    force_of_infection <- beta * I / N

    dS <- -force_of_infection * S
    dE <- force_of_infection * S - sigma * E
    dI <- sigma * E - (gamma + mu) * I
    dR <- gamma * I
    dD <- mu * I

    list(c(dS, dE, dI, dR, dD))
  })
}

# Initial conditions
initial_state <- c(
  S = 999999,
  E = 0,
  I = 1,
  R = 0,
  D = 0
)

# Time grid
times <- seq(0, 180, by = 1)

# Parameters
params <- c(
  beta = 0.55,
  sigma = 0.35,
  gamma = 0.15,
  mu = 0.08
)

# Solve the model
solution <- ode(
  y = initial_state,
  times = times,
  func = ebola_model,
  parms = params
)

solution <- as.data.frame(solution)

# Plot the results
matplot(
  solution$time,
  solution[, c("S", "E", "I", "R", "D")],
  type = "l",
  lty = 1,
  lwd = 2,
  col = c("blue", "orange", "red", "green", "black"),
  xlab = "Time (days)",
  ylab = "Population",
  main = "Ebola SEIRD Compartmental Model"
)
legend(
  "right",
  legend = c("Susceptible", "Exposed", "Infectious", "Recovered", "Deceased"),
  col = c("blue", "orange", "red", "green", "black"),
  lty = 1,
  lwd = 2
)

# Print the final state
cat("\nFinal state after 180 days:\n")
print(tail(solution, 1))

## PROMPT: Add code for a model that incorporates very specific aspects of disease spread of Ebola 

# A more detailed Ebola-specific model with:
# - community transmission
# - hospital transmission
# - funeral/burial transmission
# - recovery and death outcomes

ebola_specific_model <- function(t, state, parameters) {
  with(as.list(c(state, parameters)), {
    N <- S + E + I + H + F + R + D

    lambda_c <- beta_c * I / N
    lambda_h <- beta_h * H / N
    lambda_f <- beta_f * F / N
    lambda <- lambda_c + lambda_h + lambda_f

    dS <- -lambda * S
    dE <- lambda * S - sigma * E
    dI <- sigma * E - (alpha_h + alpha_d + gamma_i) * I
    dH <- alpha_h * I - (alpha_f + gamma_h + delta_h) * H
    dF <- alpha_f * H - (gamma_f + delta_f) * F
    dR <- gamma_i * I + gamma_h * H + gamma_f * F
    dD <- alpha_d * I + delta_h * H + delta_f * F

    list(c(dS, dE, dI, dH, dF, dR, dD))
  })
}

initial_state_specific <- c(
  S = 999999,
  E = 0,
  I = 1,
  H = 0,
  F = 0,
  R = 0,
  D = 0
)

times_specific <- seq(0, 180, by = 1)

params_specific <- c(
  beta_c = 0.45,
  beta_h = 0.25,
  beta_f = 0.60,
  sigma = 0.25,
  alpha_h = 0.20,
  alpha_d = 0.05,
  alpha_f = 0.10,
  gamma_i = 0.08,
  gamma_h = 0.12,
  gamma_f = 0.03,
  delta_h = 0.06,
  delta_f = 0.15
)

solution_specific <- ode(
  y = initial_state_specific,
  times = times_specific,
  func = ebola_specific_model,
  parms = params_specific
)

solution_specific <- as.data.frame(solution_specific)

matplot(
  solution_specific$time,
  solution_specific[, c("S", "E", "I", "H", "F", "R", "D")],
  type = "l",
  lty = 1,
  lwd = 2,
  col = c("blue", "orange", "red", "purple", "brown", "green", "black"),
  xlab = "Time (days)",
  ylab = "Population",
  main = "Detailed Ebola Model"
)
legend(
  "right",
  legend = c("Susceptible", "Exposed", "Infectious", "Hospitalized", "Funeral", "Recovered", "Deceased"),
  col = c("blue", "orange", "red", "purple", "brown", "green", "black"),
  lty = 1,
  lwd = 2
)

cat("\nFinal state for detailed Ebola model after 180 days:\n")
print(tail(solution_specific, 1))

File: VSCode/ebola_compartmental.py
## PROMPT: Code up a compartmental model for the disease spread of Ebola using Python in this file 
## INTERFACE: VS Code

"""Ebola compartmental model in Python.

This script implements an Ebola-inspired SEIHRD model with:
- S: susceptible
- E: exposed
- I: infectious in the community
- H: hospitalized
- R: recovered
- D: dead bodies awaiting burial
- B: buried bodies

The model uses a simple ordinary differential equation system and a basic
Euler solver so it can run without external scientific packages.
"""

from __future__ import annotations

import math
from typing import Dict, List, Tuple

try:
    import matplotlib.pyplot as plt
except ImportError:  # pragma: no cover - optional dependency
    plt = None


def ebola_model(t: float, y: List[float], params: Dict[str, float]) -> List[float]:
    """Return derivatives for the Ebola compartmental model."""
    S, E, I, H, R, D, B = y

    N = S + E + I + H + R + D + B
    if N <= 0:
        raise ValueError("Population size became non-positive.")

    beta_i = params["beta_i"]
    beta_h = params["beta_h"]
    beta_d = params["beta_d"]
    sigma = params["sigma"]
    rho = params["rho"]
    gamma_i = params["gamma_i"]
    gamma_h = params["gamma_h"]
    mu_i = params["mu_i"]
    mu_h = params["mu_h"]
    psi = params["psi"]

    force_of_infection = (beta_i * I + beta_h * H + beta_d * D) / N

    dS = -force_of_infection * S
    dE = force_of_infection * S - sigma * E
    dI = sigma * E - (rho + gamma_i + mu_i) * I
    dH = rho * I - (gamma_h + mu_h) * H
    dR = gamma_i * I + gamma_h * H
    dD = mu_i * I + mu_h * H - psi * D
    dB = psi * D

    return [dS, dE, dI, dH, dR, dD, dB]


def solve_model(
    params: Dict[str, float],
    initial_state: List[float],
    days: int = 180,
    dt: float = 0.1,
) -> Tuple[List[float], List[List[float]]]:
    """Solve the ODE system using a simple explicit Euler method."""
    steps = int(math.ceil(days / dt))
    times = [i * dt for i in range(steps + 1)]
    states: List[List[float]] = [[0.0] * len(initial_state) for _ in range(steps + 1)]
    states[0] = list(initial_state)

    for i in range(steps):
        derivative = ebola_model(times[i], states[i], params)
        states[i + 1] = [
            states[i][j] + dt * derivative[j]
            for j in range(len(initial_state))
        ]

    return times, states


def plot_results(times: List[float], states: List[List[float]]) -> None:
    """Plot the model compartments over time."""
    if plt is None:
        print("matplotlib is not installed; skipping plot output.")
        return

    labels = ["S", "E", "I", "H", "R", "D", "B"]
    colors = ["tab:blue", "tab:orange", "tab:red", "tab:purple", "tab:green", "black", "gray"]

    plt.figure(figsize=(10, 6))
    for idx, label in enumerate(labels):
        values = [state[idx] for state in states]
        plt.plot(times, values, label=label, color=colors[idx], linewidth=1.8)

    plt.xlabel("Time (days)")
    plt.ylabel("Population")
    plt.title("Ebola compartmental model")
    plt.grid(alpha=0.3)
    plt.legend(loc="best")
    plt.tight_layout()
    plt.show()


def main() -> None:
    params = {
        "beta_i": 0.45,
        "beta_h": 0.15,
        "beta_d": 0.10,
        "sigma": 0.20,
        "rho": 0.15,
        "gamma_i": 0.06,
        "gamma_h": 0.08,
        "mu_i": 0.03,
        "mu_h": 0.06,
        "psi": 0.12,
    }

    initial_state = [999_999, 0, 1, 0, 0, 0, 0]

    times, states = solve_model(params, initial_state, days=180, dt=0.1)

    labels = ["S", "E", "I", "H", "R", "D", "B"]
    final_state = dict(zip(labels, states[-1]))

    print("Final state after 180 days:")
    for name, value in final_state.items():
        print(f"  {name}: {value:,.0f}")

    print("\nPeak infectious count:")
    infectious_values = [state[2] for state in states]
    peak_infectious = max(infectious_values)
    peak_day = times[infectious_values.index(peak_infectious)]
    print(f"  {peak_infectious:,.0f} on day {peak_day:.1f}")

    plot_results(times, states)


if __name__ == "__main__":
    main()
